### PR TITLE
EOS-21341: Enable filter and sorting on audit logs

### PR DESCRIPTION
**Pull request by soniyamoholkar**
# Backend

# Problem Statement

- _Overview: JIRA number/GitHub Issue added to PR: [Y/N]:_ EOS-21341
- _Describe the problem this patch intends to solve._ Following problems this branch will address:
- Sorting on text fields
- Filters
- Schema to create custom table on UI

## Design

- _For Bug describe the fix here._
- _For feature, post the link to the solution page._ Links to confluence pages
- https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/494600216/Rsyslog+and+Elasticsearch
- https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/486114207/Elasticsearch-+Sorting+on+text+fields
- _Any northbound interface change and has been notified to other gatekeepers [Y/N]:_ Y

## Coding

- _Coding conventions are followed and code is consistent. Yes/No?_ Yes
- _Confirm All CODACY errors are resolved. Yes/No?_Yes

## Testing

- _Confirm that Test Cases are added (for both the cases, fix and feature). Yes/No?_ Yes
- _Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability. Yes/No?_ Yes
- _Confirm Testing was performed with installed RPM. Yes/No?_ Yes
- http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/cortx-manager/643/console
- 
image

- _Confirm all the ut/regression/smoke test has been performed [Y/N]_ Yes

## Checklist

- _PR is self-reviewed? Yes/No?_ Yes
- _GitHub Issue is updated_
- _Jira is updated_ Yes
  -  _Check if the description is clear and explained._ 
    -  _Check Acceptance Criterion is defined._
      -  _All the tests performed should be mentioned before Resolving a JIRA._
        -  _Verification needs to be done before marked as Closed/Verified_
        - _Any interface change Yes/No?_
        - _Change notified to other gatekeepers: Yes/No?_
        - _Code reviews done and addressed: Yes/No?_
        - _Codacy issues reviewed and addressed: Yes/No?_
        - _Deployment test : Done/Not?_
        - _UT/smoke test: Done/Not?_
        - _Jira number added to PR: Yes/No?_
        - _Github merge done using UI: Yes/No?_
        - _Side effects on other features - deployment/update/node-replacement_
        - _Changes to quick-start-guide/community impact_
        - _All dependent component code PR are raised or already merged Yes/No_
        - _All dependent code already merged in landing branch Yes/No_

### CODE REVIEW

File: csm/common/filter.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+from cortx.utils.data.access.filters import Compare, And, Or
+from schematics.exceptions import ConversionError
+from csm.core.blogic.models import CsmModel
+import urllib.parse
+
+class Filter():
```
Comment (pawankumarsrivastava) [inline csm/common/filter.py:21]: class Filter() --> class Filter

File: csm/common/filter.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    """Query to filter_object builder"""
+    @staticmethod
+    def parse_query(fields_of_filter):
```
Comment (pawankumarsrivastava) [inline csm/common/filter.py:24]: parse_query --> _parse_query

Comment (pawankumarsrivastava) [reply csm/common/filter.py:24]: fields_of_filter -> fields

File: csm/common/filter.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        for key in fields_of_filter:
+            if key:
+                if key == "OR" or key == "AND":
+                    operations.append(key)
+                else:
+                    key_val_fields.append(map(str.strip, key.split('=', 1)))
+        return dict(key_val_fields), operations
+
+    @staticmethod
+    def validate_query_fields(query_fields, model):
```
Comment (pawankumarsrivastava) [inline csm/common/filter.py:42]: validate_query_fields --> _validate_query_fields. Remove static.

query_fields --> fields

File: csm/common/filter.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Validate and provide default values for those keys whose values are invalid for given field.
+        :param query_fields: collection of key value pairs received as input
+        :param Type[BaseModel] model: model for constructing data mapping for index
+        :returns: Updated key_value pairs
+        """
+        default_val = model()
+        default_val_dict = default_val.to_native()
```
Comment (pawankumarsrivastava) [inline csm/common/filter.py:50]: default_val = model() ---> default_schema = model().to_native()

File: csm/common/filter.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        for key, value in query_fields.items():
+            field = eval(f"model.{key}")
+            try:
+                valid_operand = field.to_native(value)
+            except ConversionError as e:
+                updated_key_val[key] = default_val_dict[key]
+        return updated_key_val
+
+    @staticmethod
+    def _prepare_filters(query_rcvd, model):
```
Comment (pawankumarsrivastava) [inline csm/common/filter.py:61]: _prepare_filters --> prepare_filters .

Remove staticmethod

File: csm/common/filter.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        query = [urllib.parse.unquote(query_rcvd)]
+        query_fields, operations = Filter.parse_query(list(query[0][1:-1].split(" ")))
+        updated_fields = Filter.validate_query_fields(query_fields, model)
+        query_fields.update(updated_fields)
+        db_conditions = []
+        nested_operations = 0
+        filter_obj = []
+        for key, value in query_fields.items():
+            if not (db_conditions):
```
Comment (pawankumarsrivastava) [inline csm/common/filter.py:76]: Remove duplicate code. Use len

File: csm/common/filter.py (hunk 5 above)
Comment (pawankumarsrivastava) [reply csm/common/filter.py:61]: query_rcvd -> query

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/common/filter.py (hunk 1 above)
Comment (Pranali04796) [reply csm/common/filter.py]: Comment addressed

File: csm/common/filter.py (hunk 6 above)
Comment (Pranali04796) [reply csm/common/filter.py]: I have addressed it in following revision of pull request: [582](https://github.com/Seagate/cortx-manager/pull/582)

File: csm/common/filter.py (hunk 5 above)
Comment (Pranali04796) [reply csm/common/filter.py]: Addressed it by renaming variables.

File: csm/common/filter.py (hunk 2 above)
Comment (Pranali04796) [reply csm/common/filter.py]: Comment addressed by renaming.

File: csm/common/filter.py (hunk 3 above)
Comment (Pranali04796) [reply csm/common/filter.py]: Addressed it by renaming variables.

File: csm/common/filter.py (hunk 4 above)
Comment (Pranali04796) [reply csm/common/filter.py]: Reduced code as per given suggestion.

**Review by pawankumarsrivastava — APPROVED** [review]

File: csm/common/filter.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        :returns: Updated key_value pairs
+        """
+        default_val_dict = model().to_native()
+        updated_key_val = {}
+        for key, value in fields.items():
+            try:
+                field = eval(f"model.{key}")
+                valid_operand = field.to_native(value)
+            except AttributeError as err:
+                raise InvalidRequest(f"{key} not found: {err}")
```
Comment (pawankumarsrivastava) [inline csm/common/filter.py:56]: error should not have internal details like CSMAuditLogModel

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/common/filter.py (hunk 7 above)
Comment (Pranali04796) [reply csm/common/filter.py]: I have addressed it in following revision of pull request: [583](https://github.com/Seagate/cortx-manager/pull/583)

**Review by pawankumarsrivastava — APPROVED** [review]

**Pull request merged**